vivisect/parsers/elf.py

```diff
@@ -231,119 +351,362 @@ def loadElfIntoWorkspace(vw, elf, filename=None, baseaddr=None):
+                    exps = [exp for exp in vw.getExports() if exp[0] == sva]
+                    logger.debug('skipping Symbol naming for existing Import/Export: 0x%x (%r) (%r) (%r)', sva, s.name, imps, exps)
+                else:
+                    vw.makeName(sva, dmglname, filelocal=True, makeuniq=True)
+
+            except Exception as e:
+                logger.warn("WARNING:\t%r",e)
+
+        if s.st_info == Elf.STT_FUNC:
+            new_functions.append(("STT_FUNC", sva))
+
+    if addbase:
+        eentry = baseaddr + elf.e_entry
+    else:
+        eentry = elf.e_entry
 
-    if vw.isValidPointer(elf.e_entry):
-        vw.addExport(elf.e_entry, EXP_FUNCTION, '__entry', fname)
-        vw.addEntryPoint(elf.e_entry)
+    if vw.isValidPointer(eentry):
+        vw.addExport(eentry, EXP_FUNCTION, '__entry', fname)
+        new_functions.append(("ELF Entry", eentry))
 
     if vw.isValidPointer(baseaddr):
         vw.makeStructure(baseaddr, "elf.Elf32")
 
+    # mark all the entry points for analysis later
+    for cmnt, fva in new_functions:
+        logger.info('adding function from ELF metadata: 0x%x (%s)', fva, cmnt)
+        vw.addEntryPoint(fva)   # addEntryPoint queue's code analysis for later in the analysis pass
+
+    # mark all the pointers for analysis later
+    for va, tva, pname in new_pointers:
+        logger.info('adding pointer 0x%x -> 0x%x', va, tva)
+        vw.setVaSetRow('PointersFromFile', (va, tva, fname, pname))
+
     return fname
+
+def applyRelocs(elf, vw, addbase=False, baseaddr=0):
+    # process relocations / strings (relocs use Dynamic Symbols)
+    arch = arch_names.get(elf.e_machine)
+    relocs = elf.getRelocs()
+    logger.debug("reloc len: %d", len(relocs))
+    for r in relocs:
+        rtype = Elf.getRelocType(r.r_info)
+        rlva = r.r_offset
+        if addbase:
+            rlva += baseaddr
+        try:
+            # If it has a name, it's an externally
+            # resolved "import" entry, otherwise, just a regular reloc
+            name = r.getName()
+            dmglname = demangle(name)
+            logger.debug('relocs: 0x%x: %r  (%r)', rlva, dmglname, name)
+            if arch in ('i386', 'amd64'):
+                if name:
+                    if rtype == Elf.R_X86_64_IRELATIVE:
+                        # before making import, let's fix up the pointer as a BASEPTR Relocation
+                        ptr = r.r_addend
+                        vw.addRelocation(rlva, vivisect.RTYPE_BASEPTR, ptr)
+                        logger.info('Reloc: R_X86_64_IRELATIVE 0x%x', rlva)
+
+                    if rtype in (Elf.R_386_JMP_SLOT, Elf.R_X86_64_GLOB_DAT, Elf.R_X86_64_IRELATIVE):
+                        logger.info('Reloc: making Import 0x%x (name: %s/%s) ', rlva, name, dmglname)
+                        vw.makeImport(rlva, "*", dmglname)
+                        vw.setComment(rlva, name)
+
+                    elif rtype in (Elf.R_386_32, Elf.R_386_COPY):
```

Comment:
Just so I said this...somewhere, there is an ARM version of this (Elf.R_ARM_COPY) that we don't currently handle.

Though support for that may be in your ARM branch?

Reply:
correct.  let's get ARM parsing, then let's make sure the ELF parsing gets done well and quietly.  this is just to reduce noise for relocs we know we don't want to handle.